### ISSUE TITLE
Mapping - Milestone D

### ISSUE THREAD

**Issue by bsaiken** [FWAJL/FieldWorkAssistantMVC#523]

**Comment by bsaiken** [comment]
@damirius 
FYI.  I was looking at the current code on the feature_mapping branch, and the URL /map/taskLocations is displaying the map for the URL /map/currentProjectLocations.

**Comment by damirius** [comment]
@bsaiken

Can you send me screenshot of that? It looks fine on local and on dev.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

Like we agreed I'll split milestones in few parts so that they will be easier to review.
This one is for first part of Milestone D, project info window.

**Comment by FWAJL** [comment]
@damirius @bsaiken 

Thank you for the work. We'll look at it.

**Comment by bsaiken** [comment]
@damirius @FWAJL 
I have just a couple minor comments.  

Clicking on the map icon activates the info popup.  Can you also enable the popup with the sidebar icons?

When I click "Edit Boundary", the map shows the boundary, but I still need two clicks to edit the boundary: 1) click on sidebar control element "map-info-shape-icon", then 2) click on the map control element with the title "Draw a shape".  Can we skip these steps so the boundary is ready to edit directly from the link?

And, related to the item above: The map elements title="Draw a shape" and title="Stop drawing" are duplicates of the sidebar control icons.  It just creates an extra click event for each step.  Can we remove one or the other?  This would apply to all sidebar control icons.

We talked about formatting earlier.  The sidebar control icons are too large.  I prefer the icon size of the map control icons (title="Draw a shape" and title="Stop drawing").  Perhaps that means we don't use the sidebar control icons and just use the map control icons.  I don't know the coding consequences of my comments, so you will need to decide what is best.

**Comment by damirius** [comment]
@bsaiken

Biggest problem is the ruler, since it uses the same tools for drawing as boundary. So we will need to know if user wants to add/edit boundary or just measure area or distance. I use buttons to distinguish those 2 options since that google's drawing library doesn't contain separate ruler controls. 
It is possible to add marker control there since it is part of the drawing controls, but I'll need to do some rewrites for adding markers from the sidebar (for those without coordinates).

**Comment by bsaiken** [comment]
@damirius 
I thought the tools might be standard and readily available.  I have used Google My Maps and the tools don't have these conflicts.  I don't want to create a bunch of custom work.  I will share a couple screen shots on eLance that shows the tools on one of my maps as an illustration, because I like the way the tools work.

**Comment by damirius** [comment]
@bsaiken

Here is the drawing tools example: https://developers.google.com/maps/documentation/javascript/examples/drawing-tools

Those are all controls that come with that library. I use that together with geometry library that helps me with measurements for the ruler.

**Comment by damirius** [comment]
@bsaiken

I've changed few things. First off, marker highlighting is added. When you mouse over markers on map linked markers in the list will be highlighted and vice versa. 
I've edited that blue dot marker for inactive locations to be 32x32px icon just like others, so they line up better in the list.
I've also added default activated control, which we can set through query param "active". 
For example map/currentProject?active=shape will automatically open drawing controls and show path node for existing boundary.

**Comment by bsaiken** [comment]
@damirius 
From the controls you referenced above, we only need is "Add a marker", "Draw a shape" and "Stop Drawing".
We are adding the following function: when we add a new marker (Location), we get a popup box asking for the marker name.

Another control that we have, which is really different from adding a marker, is placing a sidebar marker on the map.  I didn't really say in Milestone C how to make this happen.  What we have now is using the same control icon ("Add a marker") to enable mapping an existing Location.  However, this might be better (and this works only for Locations without coordinates):
  * on mousedown, allow the user to drag the marker from the sidebar onto the map.
  * on mouseup, set the Location coordinates. 
This is kind of similar to the streetview function, where I click on the streetview icon and if I hold down the mouse and drag it shows me the street view of the final coordinates.

If the user does not drag the marker off of the sidebar, the info window will still pop up. 

Another option is that clicking a sidebar marker (one without coordinates) will drop the marker onto the map with the next click.  If the next click is not in the map area, then dropping is disabled. 

The "Draw a shape" control is used exclusively to make a Facility boundary.

The drawing tools you pointed to does not have measurement control.

**Comment by bsaiken** [comment]
@damirius 
Notwithstanding my prior comment, which are changes that can be made later, this looks good.
Info window working in Project Mapping mode.  Do you want Jeremie to review code before doing Location and Task mapping mode info windows?

**Comment by bsaiken** [comment]
@FWAJL 
Jeremie,
Go ahead and review the code.  If the info windows for the Project Map are correct, I think it's a safe bet the others will be.  That way, I can get Damir to finish this job over the weekend (provided he has time).

**Comment by FWAJL** [comment]
@damirius @bsaiken 

Just  two little comments:
1) Please use resource files to store the text values found the files. You need to create in the map resources file the resx keys for each text value and replace the current text values with $resx[“key”].
  - FieldWorkAssistantMVC/Applications/PMTool/Views/Map/Modules/popup_prompt.php at lines 24 and 25 (“Ok” and “Cancel”)
  - FieldWorkAssistantMVC/Applications/PMTool/Views/Map/Modules/popup_project_info.php at lines 17, 29, 30, 33, 34

2) When naming variables, try to be explicit. For ex, in the commit above, I renamed a variable `$items`into `$projectLocationMarkers`. I see you did rename the `$testVar` into something explicit. That is good and I encourage you to do like this.

I will let Brian decide if I can release the code with the resources as they are but keep in mind all views 

Thank you for your work.

**Comment by FWAJL** [comment]
@damirius 

I have updated my comment above (underlined part) as I missed some words. Sorry for that.

**Comment by damirius** [comment]
@FWAJL

I've pushed new commit for this. There were few problems caused by renaming `$item` variable, I've fixed that. Also I've added new resx strings for me to use with those prompts. I wanted to ask you about that and totally forgot. I noticed that some prompt titles and field names were being populated through javascript which fetched them through hidden fields which were created by looping through vars sent by the controller. Since I use only alert message prompt which is a shared module, I added those `$resx` strings directly in modules and in map.en.xml, seems easier that way since they aren't shared with other controllers.

Also I've added some alert popups for invalid coordinates since we don't want to let user enter whatever in those fields for lat and lng.

Do you want me to set this issue to code review or do we want to go through function review maybe because those alerts and what not?

Thanks for your help it is hugely appreciated!

**Comment by FWAJL** [comment]
@damirius 

Let's have @bsaiken have a look at it again. I won't have time to work on this until Monday the earliest.

Thank you for your work.

**Comment by bsaiken** [comment]
@FWAJL @damirius 
I would like to get the code for Milestone D (and E) released asap.  I just checked feature mapping branch.  Function is good.  This part is ready for code review.  The info windows sill needs to be applied to Locations (only applied to Facilities at this time).  Also, please move on to Milestone E, as it is a minor Issue, then Jeremie could do a code review Monday and we'd be done with this phase of mapping.

Thanks for adding data validation.  I see that the resource strings are set up in tooltipsandpopupstringsI, which is the right spot.  :+1  I have set up Issue #444 for data validation.  Currently we have no almost no data validation and I was going to do a complete review of the Project Manager interface once it was complete and make a comprehensive list to be done under #444.  So you gave us a jump-start.

**Comment by damirius** [comment]
@bsaiken

If that's the case then I'll work on everything left through the weekend. 
Just wanted to clarify something with you about Milestone E. It says there that when user adds address and google's geocoding api doesn't return any coordinates that it should fall back to less specific address and user should be able to "Move the marker". Now as far as I know adding new projects/facilities is the part of project/showForm?mode=add page. There isn't any map there of course, so that means that we have to redirect the user to "Selected Project" (map/currentProject) after that and give him the option to move facility marker, which shouldn't be possible after initial setup.

If you have any different solution to this, I would be glad to hear it.

Thanks

**Comment by bsaiken** [comment]
@damirius 
Thank you for asking, because a clarification is required.  Creating Facility coordinates (from the address) and moving the Facility marker are completely independent.  Generating coordinates from the Facility address must happen first, when a Project is first created.  So, there is no chance that there can be a call to map the Facility until the Facility has valid coordinates.  

Step 1:  validate the address.  If this is successful, then generating the coordinates is easy.
Step 2: if the address is invalid (according to Google),  ask the user if the address is correct.  If they respond yes, then go to step 3.  This is a possible, even likely scenario in my line of work.  Note, save the address into the database at facility.facility_address even if it is invalid.
Step 3: try to create a valid address by being less specific.  Use only the city and state from the address that was provided.  You might need a regular expression to extract the city and state.  Or, perhaps better, we could just use a popup prompt box asking for the City, State.  YOu will not save this entry into the database, you are just trying to get coordinates here.  If this works, use the coordinates that Google provides for the City.  In the screen show below I entered Meas, AZ and I see the coordinates in the URL.
Step 4: If step 3 fails, just use the USA center position.  This is in the file appsetting.xml, lines 23 and 24.

This all happens prior to the Project being created, so Facilities */must*/ have coordinates.   However, these coordinates are temporal and can be changed  the user moves the marker.  The address would not change, but the coordinates would.  

This script only runs if the Facility does not have coordinates (and this would happen only if there is a new Project being created).  If a user tries to create a copy of a Project, the new Project would already have coordinates, so the script would not run.  If the user edits the Facility address, the Facility already has coordinates, so this script would not run.  If the user tries to edit the coordinates in the /project/showForm module, then we just check that the revised coordinates are valid.  If the user enters valid coordinates that are nowhere near the Facility address that's their problem.

![screen shot 2015-03-06 at 9 58 30 am](https://cloud.githubusercontent.com/assets/8228624/6529978/30c21284-c3ea-11e4-8015-ac25fb5ab869.png)

**Comment by damirius** [comment]
@bsaiken

I have new question regarding Milestone D. I'm finishing up those info windows for locations and task locations. I've noticed one weird thing here.
In project Location info window this is in the list:
The Project name (project.project_name)
It's stated that everything should be editable, if I would take that literally it would mean that editing that field would cause whole project to be edited, not only for this Location but for all of them since project_name is stored in the project table. Now when I started working I didn't thought that this is the case because in most cases you never edit the parent element from child element, especially since we already have info window for editing project fields on the map (project info window).
I added dropdown with all available projects for that project manager so that you can switch projects for selected location.
Now, task locations have both project_name and task_name so it's the same thing for them, do we edit the whole project and whole task from there or did you meant to show dropdown select so that user can easily link that location to other project or other task?

Thanks!

**Comment by bsaiken** [comment]
@damirius 
That is a good catch.  Here is what should be done.
Display the Project name, Facility name and Task name as a breadcrumb (not editable).  I may decide, in the future, to add an href to the breadcrumb elements, for navigation.  However, the main purpose is to remind the user which Project, Facility and Task they are working on.  

In my work, Location names might be the same from Project to Project.  For example, the Location name "MW-1" is very common.  It stands for "Monitoring Well #1".  I would bet 50% of my Projects have an MW-1.

The dropdown lists need to be removed.  Moving a Location from Project to Project never happens.  And, if you did, there could be duplicate names, which would be very confusing.

**Comment by bsaiken** [comment]
@damirius 
I was just working on the feature _mapping branch and I see you are behind the dev branch.  Make sure you are doing regular git pulls otherwise we may run into conflicts.

**Comment by damirius** [comment]
@bsaiken

Thanks for the answer.

As far as the pulling goes, I'm doing that quite often and always before the commits, but as far as I know FWAJL is making the pull merges from dev to feature branches. I think I would be notified if I would try to push on the remote branch if that branch was ahead in commits. Is there something I missed?

**Comment by bsaiken** [comment]
@damirius 
This was just an FYI.  I review all the front-end code, and do some of the coding myself, so I am always ahead of what has been released to the live site.  You are probably fine.  Thanks

**Comment by damirius** [comment]
@bsaiken @FWAJL 

There's a lot of stuff in this push, I'll try to go over them here.

There are new info windows for both Project Locations and Task Locations. Both of them are active on their maps respectively. I've added breadcrumb for all infowindows including project info window like we discussed it and removed project_name, task_name fields from project location/ task location windows.
Also as mentioned earlier, clicking on location rows in the sidebar will open up those info windows just like clicking the markers on the map, except for locations with no coordinates.
`FileLoader` class is added for displaying uploaded documents, I've also changed `FileUploader` class since it had some problems with checking directories, it would try to create each time even if they exist. I've reused upload module for it, but I opted to split it from display lists. For that I had to update dropzone.js and add new stylesheet for it since we didn't include that at first. For displaying photos I opted for lightbox, I've added stylesheet and scripts. 

Note that I've removed document upload from technician form since js code had to be changed, we can't use url parsing as our method of extracting lookup information needed and store it like that in the database. For example go on dev or some other branch on technician form: `/technician/showForm?mode=edit&technician_id=45`, change the url to this `/technician/showForm?mode=edit&randomstring_id=23&technician_id=45` it will store `randomstring_id` in the `document_category` field in the db. In any case if we start working on document handling we can see what we can do about this.
Like I said you can click `View Photos` in the Project Location info window, to open up a lightbox with all photographs for that location. Please note that uploader isn't checking for valid image extensions or performing any other checks regarding validation of the file input so go easy on it.

Adding new projects will check for coordinates of entered facility address, like we discussed it will open up a prompt box asking for a city and a country name if it doesn't find them at first, after that if google's geocoding api still returns 0 results it will store default values specified in the config as the coordinates for that facility and the user will get alert box informing him that he can edit the facility coordinates on the project page. Editing existing facilities will make a basic js validation check for coordinate format, but it will not update them if you change facility address. You can still edit the coordinates manually or use the project info window for that.

Going back to All Projects or Selected Project page you will notice that markers are draggable. You can move them around and then click on them to open info window, you'll notice `Load coordinates from the map marker` option. If you move your marker and then click on that link, lat and lng fields will be updated with the new coordinates from the map, if you click `update` it will save new information.

**Comment by bsaiken** [comment]
@damirius 
There are a few issues with the info windows.
1) I do not understand what "Load coordinates from the map marker" is supposed to do.  In my test, it did not do anything.  It is OK to move the marker without any extra clicks or confirmations.  If a user moves the marker to the wrong coordinates, they can just move it back.  The purpose for my site is not to assure that we have the exact correct coordinates, just close enough to find.   So even an accidental marker move of a short distance is not worth protecting against.  The Location Markers are editable in this way.  It's OK for Projects, too.
2) It seems the control icons in the sidebar are not necessary.  Just use the control icons in the top center of the map and make all options always visible.  This also gets rid of the problem of these being too big.
3) The photos for Locations upload just fine.  We need thumbnails or a list or something so that we know there are photos present.
4) There is a strange behavior in the Task view.  If I click on a marker and try to drag it, it does not move.  But, if the final coordinates of the drag are within the Facility boundary, the Facility info window pops up.  This happens in Task View, but not Locations view.
5) "clicking on location rows in the sidebar will open up those info windows just like clicking the markers on the map".  I did not observe this behavior.
6) I would like a mouseover of the sidebar marker (& location row) to highlight the marker on the map - just like a mouseover on the map highlights the sidebar.

**Comment by damirius** [comment]
@bsaiken
First off it will be great if you can check out this video: http://irenicus.net/mapping.avi

1) On maps with facility markers, All Projects or Selected Project, you can move the marker wherever you want, when you are finished, you can click on it and info window will popup, clicking on `Load coordinates from the map marker` will then change Latitude and Longitude fields, click update and it will save them.
2) Like I said before, we need a toggle for the ruler control. Google Drawing library doesn't have ruler button, they have it on their maps but there isn't one in their API.
3) You can check previously uploaded photos for locations, or rather existing photos, by clicking `view photos` in location info window. Like every lightbox gallery you can go through them by clicking on the left/right end of the image. I can show them immediately before drag&drop div, but depending on the number of photos I believe that info window could be too big and/or broken.
4) Task view doesn't contain marker editing tools for adding new markers or moving existing ones. You just clicked on the boundary and project info window opened like it should.
5) video
6) video

**Comment by bsaiken** [comment]
@FWAJL @damirius 
I was unable to view the video, but the functionality is fine, so I am going to move this to code review.

**Comment by damirius** [comment]
@bsaiken

Great. I've added mapping.avi to elance too, if you couldn't access it there.

@FWAJL 

I'm aware that code review could take a while, there was a lot of things changed and added. If you have any questions about it please feel free to ask me so we can work them out.

Thanks

**Comment by bsaiken** [comment]
@damirius @FWAJL 
Damir,

The lightbox is working great - I like it.  It give us more potential functionality, too.
I tested it in Firefox, Chrome and Safari.

Any idea why it does not display in Opera?  I thought it was a cache issue, but I am pretty sure I cleared the Opera cache.

![screen shot 2015-03-09 at 8 29 16 pm](https://cloud.githubusercontent.com/assets/8228624/6568795/40d92730-c69b-11e4-8265-09efd3a70dbf.png)

**Comment by damirius** [comment]
@bsaiken

Hey, I'm here.

As far as I can see, that Info Window is for Projects/Facilities, they don't contain upload box or lightbox for displaying existing photos. It should be displayed when you click on the boundary or the facility marker, since from the screenshot I can see that you are on the Project Location map, can you tell me if you clicked on the boundary or the marker?

Thanks

**Comment by bsaiken** [comment]
@damirius 
I will add Opera to my list of Unit Testing tasks once the code is released to the live site.  I don't think any of my users would be using Opera, but I test in Opera (as many browsers as I can) to check cross browser functionality compatibility.  Usually, if something does not work in one browser, it is due to an inefficient coding style.  I asked you here because i thought this might be something you were aware of.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

Heya, 
I caught few small bugs which caused some issues for few of our features.
Marker id wasn't caught correctly in some browsers, I fixed it. This should solve issues with highlighting markers from sidebar, clicking on rows in the sidebar (opening info windows) and loading coordinates from the map.
I also noticed that in merge tooltip string for "No Photos available" alert box was lost, I added it again, so clicking view photos when there aren't any photos available should show alert box with appropriate notice message.

Thanks for your help!

**Comment by FWAJL** [comment]
@bsaiken @damirius

**Comment by FWAJL** [comment]
@damirius 

Another bug I think: when I highlight the location in the sidebar, the location on the map that is highlighted is always the same. Would you know why?

See the screenshots where the location on the map has the white circle but the location hovered on the side changes.

Thank
![screen shot 2015-03-11 at 06 41 54](https://cloud.githubusercontent.com/assets/7524388/6590967/6defad06-c7b9-11e4-9b21-5b71a9465888.png)
![screen shot 2015-03-11 at 06 41 48](https://cloud.githubusercontent.com/assets/7524388/6590968/6df25d08-c7b9-11e4-9058-e73a28decfde.png)
![screen shot 2015-03-11 at 06 41 43](https://cloud.githubusercontent.com/assets/7524388/6590969/6df5df00-c7b9-11e4-87f4-1f0ec57f2883.png)

**Comment by FWAJL** [comment]
@damirius I have a similar problem with task locations where the 3 active locations are mapped at the same location. See screenshots
![screen shot 2015-03-11 at 06 49 04](https://cloud.githubusercontent.com/assets/7524388/6591047/8e644000-c7ba-11e4-83d3-0ddc8874a742.png)
![screen shot 2015-03-11 at 06 48 46](https://cloud.githubusercontent.com/assets/7524388/6591046/8e642b7e-c7ba-11e4-81e8-3c70f99623ed.png)

.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

This should be fixed in this commit.
Thank you for your help in catching those pesky bugs.

**Comment by FWAJL** [comment]
@damirius 

I see you are using mostly "==" for equality checks. I would prefer you use "===" instead based on [this article](http://stackoverflow.com/questions/359494/does-it-matter-which-equals-operator-vs-i-use-in-javascript-comparisons).

What do you think about it?

Thank you

**Comment by damirius** [comment]
@FWAJL 

Depending on what do I want to check. Sometimes id is stored as string instead of integer, in those cases "==" is better since checking "2"===2 would return false. Both equality checks have their place in coding.

**Comment by bsaiken** [comment]
@damirius @FWAJL 
Is it true that x_id could be stored as a string?  I don't think that would ever be possible in our database structure.

**Comment by damirius** [comment]
@bsaiken @FWAJL

We were talking about JS part where IDs could be written as strings in html elements or stored in JS variables like that.

**Comment by FWAJL** [comment]
@damirius 

New bug: when I set a another project with active locations with no coordinates, I get this following issue:
![screen shot 2015-03-11 at 22 08 35](https://cloud.githubusercontent.com/assets/7524388/6607366/7f09795c-c83e-11e4-9804-595bac09d4d6.png)
![screen shot 2015-03-11 at 22 08 21](https://cloud.githubusercontent.com/assets/7524388/6607367/7f0c218e-c83e-11e4-8615-cce5b5fae299.png)

Can you help?

Thank you.

**Comment by damirius** [comment]
@FWAJL

So project with one active location with no coordinates? As far as I can see `item.noLatLng === true` is returning `false` there, that shouldn't happen.
Can you give me response from the ajax call (http://myapps.local/FieldWorkAssistantMVC/map/listCurrentProjectLocations)?
Here is mine:
`{
    "result": 1,
    "message": "The object have been read.",
    "noLatLngIcon": "..\/Web\/images\/ltblu-blank_maps.png",
    "activeIcon": "..\/Web\/images\/blu-circle_maps.png",
    "inactiveIcon": "..\/Web\/images\/measle_blue.png",
    "items": [
        {
            "noLatLng": true,
            "id": "313",
            "name": "asas",
            "active": "1"
        }
    ],
    "defaultPosition": {
        "lat": "39.027687",
        "lng": "-94.584595"
    },
    "boundary": "undeFtebgQvLb{LfeCckR}sA_vSsdGrlC",
    "facility_id": "55",
    "project_id": "54",
    "type": "location",
    "controls": {
        "markers": true,
        "shapes": true,
        "ruler": true
    },
    "activeControl": "pan"
}`

As you can see it returns `item` array with one element (one marker), but it returns `noLatLang: true` for it.

Thanks

**Comment by FWAJL** [comment]
@damirius 

Here it is:

`{"result":1,"message":"The object have been read.","noLatLngIcon":"..\/Web\/images\/ltblu-blank_maps.png","activeIcon":"..\/Web\/images\/blu-circle_maps.png","inactiveIcon":"..\/Web\/images\/measle_blue.png","items":[{"noLatLng":false,"id":"21","name":"Loc1","active":"1"},{"noLatLng":false,"id":"22","name":"Loc2","active":"1"},{"noLatLng":false,"id":"23","name":"Loc3","active":"1"}],"defaultPosition":{"lat":"37.6","lng":"-95.665"},"boundary":"","facility_id":"4","project_id":"7","type":"location","controls":{"markers":true,"shapes":true,"ruler":true},"activeControl":"pan"}`

The thing is why is trying to assign to `item.marker.id` when `item.marker` is undefined? (line 489 of map_manager.js)

**Comment by damirius** [comment]
@FWAJL 

I've added new commit that should fix this. I've used existing map helper method to see if markers have valid coordinates, that method returned `true` for locations with lat/lng set to `NULL` instead of a `false`. This shouldn't happen of course. Now in course of testing I was always adding new locations through the framework and new locations always ended up with "0.000000" values for both lat and lng if not specified differently. I presume that you added new location directly to the db since problem was occurring only when one of those 2 were set to `NULL`. In any case I've just added small check for that too.
To answer your last question, script was trying to add `item.id` value to `item.marker.id` since every `item` element object with `noLatLng: true` should have `marker` object nested inside of it and in this case we didn't generate `marker` object but `noLatLng` was set to `true`.

Thanks

**Comment by FWAJL** [comment]
@damirius 

I didn't add the locations via the DB but via the uploadList url. 
Will test and hopefully release by tomorrow morning.

Thank you.

**Comment by damirius** [comment]
@FWAJL

Ah, I always use Location->Add. Since they use same method I guess adding new locations through the form will set both lat and lng to empty strings and adding them through the list will only set name, but other fields will default to `null` or something similar.

Thanks for your help!

**Comment by bsaiken** [comment]
@FWAJL 
Am I correct that the mapping Issue are all complete?  I would like to close out the mapping Job (on eLance), Pay Damir, and get rid of the feature_mapping branch.

**Comment by FWAJL** [comment]
@bsaiken @damirius 

I'm releasing it. So unit test it and close the issue then.

Thank you.
